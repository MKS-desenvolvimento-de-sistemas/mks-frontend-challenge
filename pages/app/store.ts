import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import counterReducer from './counter/counterSlice';
import productsReducer from './state/actions/productsReducers';


export const store = configureStore({
  reducer: {
    counter: counterReducer,
    products: productsReducer
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
