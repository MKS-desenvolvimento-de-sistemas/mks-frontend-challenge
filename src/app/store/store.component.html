<div class="store">
    <h1>Store</h1>
    <div *ngIf="isLoading">Loading...</div>
    <app-product-card *ngFor="let product of products" [product]="product" (addToCartEvent)="addToCart($event)"></app-product-card>
  </div>
  