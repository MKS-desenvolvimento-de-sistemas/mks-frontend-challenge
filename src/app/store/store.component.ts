import { Component, OnInit } from '@angular/core';
import { ApiService, Product } from '..//services/api.service';

@Component({
  selector: 'app-store',
  templateUrl: './store.component.html',
  styleUrls: ['./store.component.scss']
})
export class StoreComponent implements OnInit {
  products: Product[] = [];
  isLoading = true;

  constructor(private apiService: ApiService) {}

  ngOnInit() {
    this.apiService.fetchProducts().then(products => {
      this.products = products;
      this.isLoading = false;
    });
  }

  addToCart(product: Product) {
    // Implementar a l√≥gica de adicionar ao carrinho
    console.log('Adicionar ao carrinho:', product);
  }
}
