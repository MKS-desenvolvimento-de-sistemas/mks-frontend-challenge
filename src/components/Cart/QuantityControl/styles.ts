import styled from "styled-components";

export const Container = styled.div`
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  flex: 2;

  position: relative;

  background: var(--primary-color-100);
  border: 0.3px solid var(--primary-color-200);
  border-radius: 4px;

  span {
    display: flex;
    justify-content: center;
    align-items: center;

    width: 33%;
    height: 95%;

    border: 0.2px solid var(--primary-color-200);
    border-top: 0px;
    border-bottom: 0px;

    font-size: 20px;
    color: var(--secondary-color-300);
  }

  small {
    display: none;
  }

  @media (min-width: 1240px) {
    max-width: 50px;
    span {
      font-size: 10px;
    }

    small {
      display: block;
      position: absolute;
      top: -11px;
      left: 0px;
      width: 100%;
      font-size: 6px;

      color: var(--secondary-color-300);
    }
  }
`;

interface IButtonProps {
  blocked?: boolean;
}

export const Button = styled.button<IButtonProps>`
  background: none;
  display: flex;
  justify-content: center;
  align-items: center;

  cursor: ${(props) => (props.blocked ? "not-allowed" : "pointer")};

  width: 33%;

  font-size: 24px;
  color: var(--secondary-color-300);

  @media (min-width: 1240px) {
    font-size: 15px;
  }
`;
