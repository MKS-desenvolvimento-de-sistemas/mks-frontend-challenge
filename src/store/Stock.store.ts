import {createSlice} from '@reduxjs/toolkit';

const carrinho = createSlice({
	name: 'carrinho',
	initialState: {
		counter: 0,
		cartOpen: 'none',
	},
	reducers: {
		increment(state) {
			state.counter += 1;
		},
		decrement(state) {
			state.counter -= 1;
		},
		abrirCarrinho(state) {
			state.cartOpen = 'flex';
		},
		fecharCarrinho(state) {
			state.cartOpen = 'none';
		},
	},
});

export const {
	decrement,
	increment,
	abrirCarrinho,
	fecharCarrinho,
} = carrinho.actions;
export default carrinho.reducer;
