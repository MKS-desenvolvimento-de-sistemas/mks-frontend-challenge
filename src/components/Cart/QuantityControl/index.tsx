import * as Styles from "./styles";

interface IQuantityControlProps {
  quantity: number;
  increment: () => void;
  decrement: () => void;
}

const QuantityControl = ({
  quantity,
  increment,
  decrement,
}: IQuantityControlProps) => {
  return (
    <Styles.Container>
      <small>Qtd:</small>
      <Styles.Button
        data-testid="decrement-button"
        type="button"
        title="Diminuir Quantidade"
        blocked={quantity <= 1}
        onClick={() => {
          if (quantity > 1) {
            decrement();
          }
        }}
      >
        -
      </Styles.Button>
      <span data-testid="quantity-display">{quantity}</span>
      <Styles.Button
        data-testid="increment-button"
        type="button"
        title="Aumentar Quantidade"
        onClick={increment}
      >
        +
      </Styles.Button>
    </Styles.Container>
  );
};

export default QuantityControl;
