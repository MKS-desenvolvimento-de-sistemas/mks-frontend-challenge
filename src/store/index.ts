import { createSlice } from "@reduxjs/toolkit";
import { configureStore } from "@reduxjs/toolkit";
import { ICartItem } from "../pages/Cart";
const initialState = { items: [] } as any;

const cartSlice = createSlice({
  name: "carrinho",
  initialState,
  reducers: {
    addToCart(state, action) {
      state.items.push(action.payload);
    },

    addAdditional(state, action) {
      if (!state.items.some((item) => item.id === action.payload.id)) return;
      const item = state.items.find((item) => item.id === action.payload.id);

      state.items.push(item);
    },
    decreaseItem(state, action) {
      const remainingItems = state.items.filter(
        (item) => item.id === action.payload.id
      );

      if (remainingItems.length === 1) return;

      const itemIndex = state.items.findIndex(
        (el) => el.id === action.payload.id
      );

      state.items.splice(itemIndex, 1);
    },
    deleteItem(state, action) {
      const arr = state.items.filter((item) => item.id !== action.payload.id);
      state.items = arr;
      // do {
      //   const itemIndex = state.items.findIndex(
      //     (el) => el.id === action.payload.id
      //   );

      //   state.items.splice(itemIndex, 1);
      // } while (state.items.some((item) => item.id === action.payload.id));
    },
    deleteAllItems(state) {
      state.items = [];
    },
  },
});

const store = configureStore({
  reducer: cartSlice.reducer,
});

export const cartActions = cartSlice.actions;

export default store;
