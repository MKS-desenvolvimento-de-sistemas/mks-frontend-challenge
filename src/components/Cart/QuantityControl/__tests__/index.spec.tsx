import "@testing-library/jest-dom";
import { fireEvent, render, screen } from "@testing-library/react";
import QuantityControl from "../";
import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import store from "../../../../redux/";
import { useState } from "react";

interface IQuantityControlWithStateProps {
  initialState?: number;
}

const QuantityControlWithState = ({
  initialState = 1,
}: IQuantityControlWithStateProps) => {
  const [quantity, setQuantity] = useState(initialState);

  const increment = () => {
    setQuantity(quantity + 1);
  };

  const decrement = () => {
    setQuantity(quantity - 1);
  };

  return (
    <Provider store={store}>
      <QuantityControl
        quantity={quantity}
        increment={increment}
        decrement={decrement}
      />
    </Provider>
  );
};

describe("QuantityControl Component", () => {
  it("should start with zero value by default", () => {
    render(<QuantityControlWithState />);

    const display = screen.getByTestId("quantity-display");

    expect(display).toHaveTextContent("1");
  });

  it("should render the decrement button correctly", () => {
    render(<QuantityControlWithState />);

    const decrementButton = screen.getByTestId("decrement-button");

    expect(decrementButton).toHaveTextContent("-");
    expect(decrementButton.getAttribute("type")).toEqual("button");
    expect(decrementButton.getAttribute("title")).toEqual(
      "Diminuir Quantidade"
    );
  });

  it("should render the increment button correctly", () => {
    render(<QuantityControlWithState />);

    const incrementButton = screen.getByTestId("increment-button");

    expect(incrementButton).toHaveTextContent("+");
    expect(incrementButton.getAttribute("type")).toEqual("button");
    expect(incrementButton.getAttribute("title")).toEqual(
      "Aumentar Quantidade"
    );
  });

  it("should not reduce the value if it is less than or equal to 1", () => {
    render(<QuantityControlWithState />);

    const decrementButton = screen.getByTestId("decrement-button");
    const display = screen.getByTestId("quantity-display");

    expect(display).toHaveTextContent("1");

    // QUANTITY É MENOR QUE 1 QUANDO INICIA
    expect(decrementButton).toHaveStyle({ cursor: "not-allowed" });

    fireEvent.click(decrementButton);

    // VALOR NÃO DIMINUIR, POIS NÃO É MAIOR QUE 1
    expect(display).toHaveTextContent("1");
  });

  it("should reduce the value if it is greater than 1", () => {
    render(<QuantityControlWithState initialState={3} />);

    const decrementButton = screen.getByTestId("decrement-button");
    const display = screen.getByTestId("quantity-display");

    expect(display).toHaveTextContent("3");

    expect(decrementButton).toHaveStyle({ cursor: "pointer" });

    fireEvent.click(decrementButton);

    expect(display).toHaveTextContent("2");
  });

  it("should increase the value", () => {
    render(<QuantityControlWithState />);

    const incrementButton = screen.getByTestId("increment-button");
    const display = screen.getByTestId("quantity-display");

    expect(display).toHaveTextContent("1");

    fireEvent.click(incrementButton);

    expect(display).toHaveTextContent("2");
  });

  it("should match the snapshot", () => {
    const tree = renderer.create(<QuantityControlWithState />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
