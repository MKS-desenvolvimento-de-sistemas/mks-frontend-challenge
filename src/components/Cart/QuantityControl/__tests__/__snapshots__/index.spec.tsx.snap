// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuantityControl Component should match the snapshot 1`] = `
<div
  className="sc-bczRLJ hDDUiJ"
>
  <small>
    Qtd:
  </small>
  <button
    className="sc-gsnTZi fIZqiW"
    data-testid="decrement-button"
    onClick={[Function]}
    title="Diminuir Quantidade"
    type="button"
  >
    -
  </button>
  <span
    data-testid="quantity-display"
  >
    1
  </span>
  <button
    className="sc-gsnTZi eQzbAZ"
    data-testid="increment-button"
    onClick={[Function]}
    title="Aumentar Quantidade"
    type="button"
  >
    +
  </button>
</div>
`;
